### PR TITLE
[batch] use compare and swap when updating a record

### PR DESCRIPTION
**Pull request by jigold**
Stacked on #6519

### CODE REVIEW

File: batch/batch/batch.py [hunk 1]
```diff
@@ -284,28 +288,40 @@ class Job:
...
-    async def _mark_job_task_complete(self, task_name, log, exit_code, new_state, pod_status):
-        self.exit_codes[self._task_idx] = exit_code
-        self.pod_statuses[self._task_idx] = pod_status
-
+    async def _mark_job_task_complete(self, task_name, pod_log, exit_code, new_state, pod_status):
         uri = None
-        if log is not None:
-            uri, err = await app['log_store'].write_gs_log_file(*self.id, task_name, log)
+        if pod_log is not None:
+            uri, err = await app['log_store'].write_gs_log_file(*self.id, task_name, pod_log)
```
Comment (danking) [inline batch/batch/batch.py]: needs a random number on file name in case this is the second time the task was run.

File: batch/batch/batch.py [hunk 2]
```diff
@@ -284,28 +288,40 @@ class Job:
...
                 log.info(f'job {self.full_id} will have a missing log due to {err}')
-            else:
-                self.log_uris[self._task_idx] = uri
+
+        compare_items = {'state': self._state, 'task_idx': self._task_idx}
+        n_updated = await db.jobs.update_with_log_ec(*self.id, task_name, uri, exit_code, pod_status,
+                                                     compare_items=compare_items,
+                                                     task_idx=self._task_idx,
+                                                     duration=self.duration,
+                                                     state=new_state)
```
Comment (danking) [inline batch/batch/batch.py]: verify state & uri are in expected situation

File: batch/batch/database.py [hunk 3]
```diff
@@ -89,7 +89,11 @@ def new_record_template(self, *field_names):
                     set_template = ", ".join([f'`{k.replace("`", "``")}` = %s' for k, v in set_items.items()])
                     set_values = set_items.values()
                     sql = f"UPDATE `{self.name}` SET {set_template} WHERE {where_template}"
-                    await cursor.execute(sql, (*set_values, *where_values))
+                    log.info(sql, *(*set_values, *where_values))
+                    result = await cursor.execute(sql, (*set_values, *where_values))
+                    log.info(result)
```
Comment (danking) [inline batch/batch/database.py]: rm -rf

File: batch/batch/batch.py (hunk 1 above)
Comment (jigold) [reply batch/batch/batch.py]: this is a harder change to make and not essential

**Pull request merged**